core/src/main/java/org/ldaptive/transport/DefaultOperationHandle.java

```diff
@@ -567,40 +676,55 @@ public void exception(final LdapException e)
       }
     }
     exception = e;
+    responseSemaphore.release();
     complete();
   }
 
 
   /**
    * Indicates that a protocol message was consumed by a supplied consumer.
+   *
+   * @deprecated  Use {@link #consumedMessage(Message)}
    */
+  @Deprecated
   protected void consumedMessage()
   {
     consumedMessage = true;
+    responseSemaphore.release();
+  }
+
+
+  /**
+   * Indicates that a protocol message was consumed by a supplied consumer.
+   *
+   * @param  message  that was consumed
+   */
+  protected void consumedMessage(final Message message)
+  {
+    consumedMessage = true;
+    if (responseTimeoutCondition.test(message)) {
+      responseSemaphore.release();
+    }
   }
 
 
   /**
    * Releases the latch and sets the response as received. Invokes {@link #onComplete}. Handle is considered done when
    * this is invoked.
    */
-  private void complete()
+  private synchronized void complete()
```

Comment:
I don't understand the need to make this method synchronized. The same instance variables are at play, so it seems you don't need method synchronization unless you feel you missed something previously.

Reply:
I missed something previously. There is a potential race condition where a result and an exception could occur in a timespan that would allow two separate threads to enter this method. This could potentially result in `onComplete` and/or `connection#complete` being invoked twice.

Reply:
Understood, good catch.